### PR TITLE
Npm harvester

### PR DESCRIPTION
**Pull request by JeroenKnoops**
Add Npm Harvester

### CODE REVIEW

**Comment by JeroenKnoops** [comment]
Still draft, because I think we can get more information out of this harvester.

**Comment by JeroenKnoops** [comment]
When only using the pipenv harvester (remove clearlydefined harvester) you get this amount of data per package in the spdx file:

```
## Start of package pkg:pypi/mako@1.1.4
PackageName: mako
SPDXID: SPDXRef-2
PackageVersion: 1.1.4
PackageSummary: Mako
PackageDescription: A super-fast templating language that borrows the  best ideas from the existing templating languages.
PackageHomePage: https://www.makotemplates.org/
ExternalRef: PACKAGE-MANAGER purl pkg:pypi/mako@1.1.4
PackageDownloadLocation: NOASSERTION
PackageLicenseConcluded: NOASSERTION
PackageLicenseDeclared: MIT
FilesAnalyzed: false
PackageCopyrightText: NOASSERTION
Relationship: SPDXRef-2 DYNAMIC_LINK SPDXRef-3
```

When using
 only npm you get a similar record:
```
## Start of package pkg:npm/js-tokens@4.0.0
PackageName: js-tokens
SPDXID: SPDXRef-4
PackageVersion: 4.0.0
PackageSummary: js-tokens
PackageDescription: A regex that tokenizes JavaScript.
PackageHomePage: NOASSERTION
ExternalRef: PACKAGE-MANAGER purl pkg:npm/js-tokens@4.0.0
PackageDownloadLocation: NOASSERTION
PackageLicenseConcluded: NOASSERTION
PackageLicenseDeclared: MIT
FilesAnalyzed: false
PackageCopyrightText: NOASSERTION
```

When using the clearlydefined harvester you get a lot more data:
```
## Start of package pkg:npm/js-tokens@4.0.0
PackageName: js-tokens
SPDXID: SPDXRef-4
PackageVersion: 4.0.0
PackageSummary: js-tokens
PackageDescription: A regex that tokenizes JavaScript.
PackageHomePage: https://github.com/lydell/js-tokens#readme
ExternalRef: PACKAGE-MANAGER purl pkg:npm/js-tokens@4.0.0
PackageDownloadLocation: https://registry.npmjs.com/js-tokens/-/js-tokens-4.0.0.tgz
PackageChecksum: SHA256: d884c7a2d8adb5568c1272d92b4f9c62707f4226cf9e7b22e7b957c7361e3c53
PackageChecksum: SHA1: 19203fb59991df98e3a287050d4647cdeaf32499
PackageLicenseConcluded: NOASSERTION
PackageLicenseDeclared: MIT
PackageLicenseInfoFromFiles: MIT
FilesAnalyzed: true
PackageCopyrightText: NOASSERTION
```

@timovandeput Shouldn't we get more data from both PyPi-harvester and NPM-harvester?

**Review by timovandeput — CHANGES_REQUESTED** [review]

File: src/main/java/com/philips/research/bombase/core/npm/domain/NpmClient.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+                return Optional.empty();
+            }
+            if (!response.isSuccessful()) {
+                throw new NpmException("Npm server responded with status " + response.code());
+            }
+            return Optional.ofNullable(response.body());
+        } catch (JsonProcessingException e) {
+            throw new IllegalArgumentException("JSON formatting error", e);
+        } catch (IOException e) {
+            throw new NpmException("Npm is not reachable");
```
[the remark this answers is not in the document]
Comment (JeroenKnoops) [reply src/main/java/com/philips/research/bombase/core/npm/domain/NpmClient.java:62]: I'm not sure what our strategy is with acronyms in class names. I tend to format abbreviations as a word if part of a longer class name.
So I will leave the class name as is, and change the exception.

File: src/main/java/com/philips/research/bombase/core/npm/domain/NpmAPI.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        }
+
+        @Override
+        public Optional<String> getLicense() {
+            return Optional.ofNullable(license);
+        }
+
+        @Override
+        public Optional<URI> getSourceUrl() {
+            return Optional.ofNullable(dist.tarball);
```
[the remark this answers is not in the document]
Comment (timovandeput) [reply src/main/java/com/philips/research/bombase/core/npm/domain/NpmAPI.java:55]: Removed the annotation on the field, assuming it is not an optional field in the API.

File: src/main/java/com/philips/research/bombase/core/npm/domain/NpmClient.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (timovandeput) [reply src/main/java/com/philips/research/bombase/core/npm/domain/NpmClient.java:62]: Convention in Java is to stick to all-uppercase for abbreviations of up to 2 characters (like `IOException`), and only uppercase the first character for longer abbreviations (like `JsonProcessingException`). So the name of the class is correct.

But the name in the _text_ should be "NPM", as this is the official acronym.

File: src/main/java/com/philips/research/bombase/core/npm/domain/NpmClient.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    Optional<PackageDefinition> getPackage(PackageURL purl) {
+        return query(rest.getDefinition(purl.getName(), purl.getVersion()))
+                .map(def -> (PackageDefinition) def);
+    }
+
+    private <T> Optional<T> query(Call<? extends T> query) {
+        try {
+            final var response = query.execute();
+            if (response.code() == 404) {
+                return Optional.empty();
```
[the remark this answers is not in the document]
Comment (timovandeput) [reply src/main/java/com/philips/research/bombase/core/npm/domain/NpmClient.java:53]: Added "info" logging to all harvesters when no metadata was found for a package.

**Comment by JeroenKnoops** [comment]
Closes #44

**Review by timovandeput — APPROVED** [review]

**Comment by JeroenKnoops** [comment]
On the fetching of the author we get an error in some cases:

```
(although at least one Creator exists): no String-argument constructor/factory method to deserialize from String value ('James Coglan <jcoglan@gmail.com> (http://jcoglan.com/)')
```

Full log
```
2021-07-07 17:57:04.933 ERROR 78180 --- [         pool-1] .a.i.SimpleAsyncUncaughtExceptionHandler : Unexpected exception occurred invoking async method: public void com.philips.research.bombase.core.meta.registry.QueuedTaskRunner.execute(com.github.packageurl.PackageURL,java.util.function.Consumer,java.util.function.Consumer)

java.lang.IllegalArgumentException: JSON formatting error
	at com.philips.research.bombase.core.npm.domain.NpmClient.query(NpmClient.java:60) ~[classes!/:0.1.1-SNAPSHOT]
	at com.philips.research.bombase.core.npm.domain.NpmClient.getPackage(NpmClient.java:45) ~[classes!/:0.1.1-SNAPSHOT]
	at com.philips.research.bombase.core.npm.domain.NpmHarvester.harvest(NpmHarvester.java:48) ~[classes!/:0.1.1-SNAPSHOT]
	at com.philips.research.bombase.core.npm.domain.NpmHarvester.lambda$onUpdated$0(NpmHarvester.java:44) ~[classes!/:0.1.1-SNAPSHOT]
	at com.philips.research.bombase.core.meta.registry.QueuedTaskRunner.lambda$execute$0(QueuedTaskRunner.java:38) ~[classes!/:0.1.1-SNAPSHOT]
	at java.base/java.util.Optional.ifPresent(Optional.java:176) ~[na:na]
	at com.philips.research.bombase.core.meta.registry.QueuedTaskRunner.execute(QueuedTaskRunner.java:36) ~[classes!/:0.1.1-SNAPSHOT]
	at com.philips.research.bombase.core.meta.registry.QueuedTaskRunner$$FastClassBySpringCGLIB$$b8050159.invoke(<generated>) ~[classes!/:0.1.1-SNAPSHOT]
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218) ~[spring-core-5.3.7.jar!/:5.3.7]
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:779) ~[spring-aop-5.3.7.jar!/:5.3.7]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163) ~[spring-aop-5.3.7.jar!/:5.3.7]
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:750) ~[spring-aop-5.3.7.jar!/:5.3.7]
	at org.springframework.aop.interceptor.AsyncExecutionInterceptor.lambda$invoke$0(AsyncExecutionInterceptor.java:115) ~[spring-aop-5.3.7.jar!/:5.3.7]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:832) ~[na:na]
Caused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot construct instance of `com.philips.research.bombase.core.npm.domain.NpmAPI$PersonJson` (although at least one Creator exists): no String-argument constructor/factory method to deserialize from String value ('James Coglan <jcoglan@gmail.com> (http://jcoglan.com/)')
 at [Source: (okhttp3.ResponseBody$BomAwareReader); line: 1, column: 156] (through reference chain: com.philips.research.bombase.core.npm.domain.NpmAPI$ResponseJson["author"])
	at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:63) ~[jackson-databind-2.12.3.jar!/:2.12.3]
	at com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1588) ~[jackson-databind-2.12.3.jar!/:2.12.3]
	at com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1213) ~[jackson-databind-2.12.3.jar!/:2.12.3]
	at com.fasterxml.jackson.databind.deser.std.StdDeserializer._deserializeFromString(StdDeserializer.java:311) ~[jackson-databind-2.12.3.jar!/:2.12.3]
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromString(BeanDeserializerBase.java:1495) ~[jackson-databind-2.12.3.jar!/:2.12.3]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:207) ~[jackson-databind-2.12.3.jar!/:2.12.3]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:197) ~[jackson-databind-2.12.3.jar!/:2.12.3]
	at com.fasterxml.jackson.databind.deser.impl.FieldProperty.deserializeAndSet(FieldProperty.java:138) ~[jackson-databind-2.12.3.jar!/:2.12.3]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:324) ~[jackson-databind-2.12.3.jar!/:2.12.3]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:187) ~[jackson-databind-2.12.3.jar!/:2.12.3]
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:322) ~[jackson-databind-2.12.3.jar!/:2.12.3]
	at com.fasterxml.jackson.databind.ObjectReader._bindAndClose(ObjectReader.java:2033) ~[jackson-databind-2.12.3.jar!/:2.12.3]
	at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1458) ~[jackson-databind-2.12.3.jar!/:2.12.3]
	at retrofit2.converter.jackson.JacksonResponseBodyConverter.convert(JacksonResponseBodyConverter.java:33) ~[converter-jackson-2.9.0.jar!/:na]
	at retrofit2.converter.jackson.JacksonResponseBodyConverter.convert(JacksonResponseBodyConverter.java:23) ~[converter-jackson-2.9.0.jar!/:na]
	at retrofit2.OkHttpCall.parseResponse(OkHttpCall.java:243) ~[retrofit-2.9.0.jar!/:na]
	at retrofit2.OkHttpCall.execute(OkHttpCall.java:204) ~[retrofit-2.9.0.jar!/:na]
	at com.philips.research.bombase.core.npm.domain.NpmClient.query(NpmClient.java:51) ~[classes!/:0.1.1-SNAPSHOT]
	... 16 common frames omitted
```

**Comment by JeroenKnoops** [comment]
When disabling the clearlydefined harvester, we still do not get any source information.

Part of the SPDX output.
```
## Start of package pkg:npm/type@1.2.0
PackageName: type
SPDXID: SPDXRef-985
PackageVersion: 1.2.0
PackageSummary: type
PackageDescription: Runtime validation and processing of JavaScript types
PackageHomePage: NOASSERTION
ExternalRef: PACKAGE-MANAGER purl pkg:npm/type@1.2.0
PackageDownloadLocation: NOASSERTION
PackageLicenseConcluded: NOASSERTION
PackageLicenseDeclared: ISC
FilesAnalyzed: false
PackageCopyrightText: NOASSERTION
```

**Review by AnnieJovitha — APPROVED** [review]

**Comment by JeroenKnoops** [comment]
I will rerun my tests!

**Comment by JeroenKnoops** [comment]
:shipit: 🍾

**Pull request merged**